File: backend/services/pms_service/app/schemas/sale_corrections.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

from .inventory_validation import Money, Units
from .transactions import CancelTransaction, PaymentMethod, TransactionInput, Version


class CorrectionLine(TransactionInput):
    sale_item_id: UUID
    quantity: Units


class CorrectionCreate(CancelTransaction):
    prescription_version: Version | None = None
    kind: Literal["not_collected", "customer_return"]
    stock_confirmed: Literal[True]
    items: list[CorrectionLine] = Field(min_length=1, max_length=200)

    @field_validator("items")
    @classmethod
    def unique_lines(cls, lines):
        if len({line.sale_item_id for line in lines}) != len(lines):
            raise ValueError("include each receipt line only once")
        return lines


class RefundCreate(CancelTransaction):
    amount_cents: Money = Field(gt=0)
    payment_method: PaymentMethod
    payment_ref: str = Field(min_length=3, max_length=128)
    payment_confirmed: Literal[True]


class RefundVoid(CancelTransaction):
    entry_was_incorrect: Literal[True]


class PrescriptionAllocation(TransactionInput):
    sale_item_id: UUID
    prescription_item_id: UUID


class ReconcilePrescription(CancelTransaction):
    prescription_version: Version
    allocations: list[PrescriptionAllocation] = Field(min_length=1, max_length=200)


class CorrectionItemOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    sale_item_id: UUID
    quantity: int
    credit_cents: int


class CorrectionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    sale_id: UUID
    number: str
    kind: str
    reason: str
    credit_cents: int
    actor_staff_id: UUID
    actor_name: str | None = None
    created_at: datetime
    items: list[CorrectionItemOut] = []


class RefundOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    sale_id: UUID
    number: str
    amount_cents: int
    payment_method: str
    payment_ref: str
    reason: str
    actor_staff_id: UUID
    actor_name: str | None = None
    created_at: datetime
    status: str
    void_reason: str | None
    voided_at: datetime | None
    voided_by: UUID | None


class CorrectionQuote(BaseModel):
    credit_cents: int
    currency: str
    items: list[CorrectionItemOut]


class CorrectionList(BaseModel):
    items: list[CorrectionOut]
    total: int
    limit: int
    offset: int


class RefundList(BaseModel):
    items: list[RefundOut]
    total: int
    limit: int
    offset: int

File: frontend/pms_web/src/lib/repositories/sale-corrections.ts
import api from "@/lib/api/client";
import type { Sale } from "./sales";
import type { CancelTransaction, Page, PaymentMethod } from "./transactions";

export interface CorrectionCreate extends CancelTransaction {
  prescription_version?: number;
  kind: "not_collected" | "customer_return";
  stock_confirmed: true;
  items: { sale_item_id: string; quantity: number }[];
}
export interface CorrectionQuote {
  credit_cents: number;
  currency: string;
  items: { sale_item_id: string; quantity: number; credit_cents: number }[];
}
export interface Correction extends Omit<CorrectionQuote, "currency"> {
  id: string;
  sale_id: string;
  number: string;
  kind: CorrectionCreate["kind"];
  reason: string;
  actor_staff_id: string;
  actor_name?: string | null;
  created_at: string;
}
export interface RefundCreate extends CancelTransaction {
  amount_cents: number;
  payment_method: PaymentMethod;
  payment_ref: string;
  payment_confirmed: true;
}
export interface Refund {
  id: string;
  sale_id: string;
  number: string;
  amount_cents: number;
  payment_method: string;
  payment_ref: string;
  reason: string;
  actor_staff_id: string;
  actor_name?: string | null;
  created_at: string;
  status: "recorded" | "voided";
  void_reason: string | null;
  voided_at: string | null;
  voided_by: string | null;
}
export interface ReconcilePrescription extends CancelTransaction {
  prescription_version: number;
  allocations: { sale_item_id: string; prescription_item_id: string }[];
}
const options = (key: string, signal?: AbortSignal) => ({
  signal,
  headers: { "Idempotency-Key": key },
});
export const correctionsRepo = {
  list: async (
    id: string,
    offset = 0,
    signal?: AbortSignal,
  ): Promise<Page<Correction>> =>
    (
      await api.get(`/v1/sales/${id}/corrections`, {
        params: { offset, limit: 25 },
        signal,
      })
    ).data,
  refunds: async (
    id: string,
    offset = 0,
    signal?: AbortSignal,
  ): Promise<Page<Refund>> =>
    (
      await api.get(`/v1/sales/${id}/refunds`, {
        params: { offset, limit: 25 },
        signal,
      })
    ).data,
  quote: async (
    id: string,
    body: CorrectionCreate,
    signal?: AbortSignal,
  ): Promise<CorrectionQuote> =>
    (await api.post(`/v1/sales/${id}/corrections/quote`, body, { signal }))
      .data,
  correct: async (
    id: string,
    body: CorrectionCreate,
    key: string,
    signal?: AbortSignal,
  ): Promise<Correction> =>
    (await api.post(`/v1/sales/${id}/corrections`, body, options(key, signal)))
      .data,
  refund: async (
    id: string,
    body: RefundCreate,
    key: string,
    signal?: AbortSignal,
  ): Promise<Refund> =>
    (await api.post(`/v1/sales/${id}/refunds`, body, options(key, signal)))
      .data,
  voidRefund: async (
    id: string,
    refundId: string,
    body: CancelTransaction & { entry_was_incorrect: true },
    key: string,
    signal?: AbortSignal,
  ): Promise<Refund> =>
    (
      await api.post(
        `/v1/sales/${id}/refunds/${refundId}/void`,
        body,
        options(key, signal),
      )
    ).data,
  reconcile: async (
    id: string,
    body: ReconcilePrescription,
    key: string,
    signal?: AbortSignal,
  ): Promise<Sale> =>
    (
      await api.post(
        `/v1/sales/${id}/reconcile-prescription`,
        body,
        options(key, signal),
      )
    ).data,
};
